### PR TITLE
[dot] Update tensor dot implementation

### PR DESCRIPTION
**Pull request by kparichay**
Update the implementation of tensor dot product
This update applies dot on the last dimesnion of input and second last dimension of passed tensor `m`
This allows use of dot in fc layer directly eliminating calling transpose explicitly but rather use transpose in blas

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/src/tensor.cpp [hunk 1]
```diff
@@ -592,143 +592,118 @@ Tensor Tensor::sum(int axis) const {
...
+  Tensor result;
 
-  int mwidth = m.dim.width();
-  Tensor result(dim.batch(), 1, dim.height(), mwidth);
+  unsigned int M, N, K, lda, ldb, ldc;
+
+  if (!trans && !trans_m) {
+    if (dim2 != mdim1)
+      throw std::runtime_error(
+        "Error: imcompatible dimensions for dot product");
```
Comment (zhoonit) [inline nntrainer/src/tensor.cpp:611]:
```suggestion
        "Error: incompatible dimensions for dot product");
```

And for others below

File: nntrainer/src/tensor.cpp [hunk 2]
```diff
@@ -592,143 +592,118 @@ Tensor Tensor::sum(int axis) const {
...
- * If the dim.batch() size of m is one, the it is reused for
- * every calculation along with dim.batch().
- * TODO: support dim.channel() > 1.
- */
-Tensor Tensor::dot(Tensor const &m) const {
-  if (dim.width() != m.dim.height()) {
-    throw std::runtime_error("Error dim.width() != m.dim.height()");
+Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+  if (m.dim.rank() > 2) {
+    throw std::runtime_error("Error: support only for rank of dot matrix <= 2");
```
Comment (zhoonit) [inline nntrainer/src/tensor.cpp:597]:
```suggestion
    throw exception::not_supported("Error: support only for rank of dot matrix <= 2");
```

File: nntrainer/src/tensor.cpp [hunk 3]
```diff
@@ -592,143 +592,118 @@ Tensor Tensor::sum(int axis) const {
...
-/**
- * If the dim.batch() size of m is one, the it is reused for
- * every calculation along with dim.batch().
- * TODO: support dim.channel() > 1.
- */
-Tensor Tensor::dot(Tensor const &m) const {
-  if (dim.width() != m.dim.height()) {
-    throw std::runtime_error("Error dim.width() != m.dim.height()");
+Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+  if (m.dim.rank() > 2) {
```
Comment (zhoonit) [inline nntrainer/src/tensor.cpp:596]: We discussed offline to add safeguard to `when (this->dim.rank() > 2)` in case of `trans` is true

**Comment by jijoongmoon** [comment]
I think once we merge this PR, then we need to modify little bit something like #491

File: nntrainer/src/tensor.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/src/tensor.cpp:611]: done.

File: nntrainer/src/tensor.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/tensor.cpp:597]: done.

File: nntrainer/src/tensor.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/tensor.cpp:601]: done.

**Comment by kparichay** [comment]
> I think once we merge this PR, then we need to modify little bit something like #491

Yes, once #491 is merged, can directly use gemm from there.

**Comment by kparichay** [comment]
more unit tests for dot will soon follow.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by zhoonit — APPROVED** [review]
LGTM :)

**Pull request merged**